src/modules/mc_pos_control/Takeoff/Takeoff.cpp

```diff
@@ -38,73 +38,119 @@
 #include "Takeoff.hpp"
 #include <mathlib/mathlib.h>
 
-void Takeoff::generateInitialRampValue(const float hover_thrust, float velocity_p_gain)
+struct StateParams
 {
-	velocity_p_gain = math::max(velocity_p_gain, 0.01f);
-	_takeoff_ramp_vz_init = -hover_thrust / velocity_p_gain;
-}
-
-void Takeoff::updateTakeoffState(const bool armed, const bool landed, const bool want_takeoff,
+	StateParams(Takeoff *parent, const bool armed, const bool landed, const bool want_takeoff,
 				 const float takeoff_desired_vz, const bool skip_takeoff, const hrt_abstime &now_us)
-{
-	_spoolup_time_hysteresis.set_state_and_update(armed, now_us);
-
-	switch (_takeoff_state) {
-	case TakeoffState::disarmed:
-		if (armed) {
-			_takeoff_state = TakeoffState::spoolup;
-
-		} else {
-			break;
-		}
+		: _parent(parent),
+		  _armed(armed), 
+		  _landed(landed), 
+		  _want_takeoff(want_takeoff), 
+		  _skip_takeoff(skip_takeoff), 
+		  _takeoff_desired_vz(takeoff_desired_vz), 
+		  _now_us(now_us)
+	{
+	}
 
-	// FALLTHROUGH
-	case TakeoffState::spoolup:
-		if (_spoolup_time_hysteresis.get_state()) {
-			_takeoff_state = TakeoffState::ready_for_takeoff;
+	Takeoff *_parent;
+	bool _armed, _landed, _want_takeoff, _skip_takeoff;
+	float _takeoff_desired_vz;
+	const hrt_abstime &_now_us;
+};
 
-		} else {
-			break;
-		}
+class Takeoff::StateHandler
+{
+public:
+	static void handleDisarmed(const StateParams& params);
```

Comment:
maybe these should be inlined, then we are closer to the switch fallthrough implementation